pages/docs/couple-your-code/couple-your-code-waveform.md

```diff
@@ -0,0 +1,167 @@
+
+{% include important.html content="If subcycling is used, data that was written before the last time step in the window will be ignored and will not be exchanged. This means that coupling data has a constant value (in time) within one coupling window." %}
+
+### Example for subcycling without waveform iteration
+
+The figure below visualizes this situation for a single coupling window ranging from $$t_\text{ini}$$ to $$t_\text{ini}+\Delta t$$:
+
+![Coupling data exchange without interpolation](images/docs/couple-your-code/couple-your-code-waveform/WaveformConstant.png)
+
+The two participants Dirichlet $$\mathcal{D}$$ and Neumann $$\mathcal{N}$$ use their respective time step sizes $$\delta t$$ and produce coupling data $$c$$ at the end of each time step. But only the very last samples $$c_{\mathcal{N}\text{end}}$$ and $$c_{\mathcal{D}\text{end}}$$ are exchanged. If the Dirichlet participant $$\mathcal{D}$$ calls `readBlockVectorData`, it will always receive the same value $$c_{\mathcal{N}\text{end}}$$ from the Neumann participant $$\mathcal{N}$$, independent from the current time step.
+
+## Linear interpolation in a time window
+
+A simple extension of the existing API is to apply linear interpolation inside of a time window to get smoother coupling boundary conditions. With this approach time-dependent functions (so-called *waveforms*) are exchanged between the participant. Since these waveforms are exchanged iteratively in implicit coupling, we call this procedure *waveform iteration*. Exchanging waveforms leads to a more robust subcycling and allows us to support higher order time stepping (for details, see [here](https://onlinelibrary.wiley.com/doi/full/10.1002/nme.6443)).
+
+### Example for waveform iteration with linear interpolation
+
+Linear interpolation between coupling boundary conditions of the previous and the current time window is illustrated below:
+
+![Coupling data exchange with linear interpolation](images/docs/couple-your-code/couple-your-code-waveform/WaveformLinear.png)
+
+If the Dirichlet participant $$\mathcal{D}$$ calls `readBlockVectorData`, it samples the data from a time-dependent function $$c_\mathcal{D}(t)$$. This function is created from linear interpolation of the first and the last sample $$c_{\mathcal{D}0}$$ and $$c_{\mathcal{D}5}$$ created by the Neumann participant $$\mathcal{N}$$ in the current time window. This allows $$\mathcal{D}$$ to sample the coupling condition at arbitrary times $$t$$ inside the current time window.
+
+## Experimental API for waveform iteration
+
+preCICE offers an experimental API for using linear interpolation on time windows. Here, `readBlockVectorData` accepts an additional argument `dt`. This allows us to choose the time where the interpolant should be sampled:
+
+```cpp
+// stable API with constant data in time window
+void readBlockVectorData(int dataID, int size, const int* valueIndices, double* values) const;
```

Comment:
Is the data really constant within a time window? I thought we get the linear interpolation here as well? (if the read side is subcycling).

Reply:
Yes. Data is constant, if this function call is used. From the current state of the [waveform pull request](https://github.com/BenjaminRodenberg/precice/blob/3b0a94b50f2d781019379ff285e9db8b0206e22a/src/precice/impl/SolverInterfaceImpl.hpp#L420-L436):

```
  /**
   * @brief Reads vector data values given as block from the interface mesh at the end of the time window.
   * ...
   */
  void readBlockVectorData(
      int        toDataID,
      int        size,
      const int *valueIndices,
      double *   values) const;
```

Looking [at the implementation](https://github.com/BenjaminRodenberg/precice/blob/3b0a94b50f2d781019379ff285e9db8b0206e22a/src/precice/impl/SolverInterfaceImpl.cpp#L1192-L1194) also shows this:

```
  double timeStepDt        = _couplingScheme->getThisTimeWindowRemainder(); // samples at end of time window
  bool   checkExperimental = false;
  return readBlockVectorData(dataID, size, valueIndices, timeStepDt, values, checkExperimental);
```

The reason is that when we call `read...Data(...)` without `dt`, we cannot know where the user wants to sample (even if subcycling is done). Remember `advance(dt)` defines the length of the time step. `advance` is called after `read` and therefore `read` cannot make any better assumptions than that the user want to read data at the end of the window.

Reply:
@uekerman does this sound reasonable? If yes, feel free to resolve this discussion.

Reply:
Added a clarification in 18a89043f8112ce56e8e0cf9fd73b9e5d5adb7df.

Reply:
Good point with the warning. Implemented it in https://github.com/precice/precice/pull/1029/commits/909520f82e8b5f6887941296b1a7dfcb117d67ea.